decomposer/neuro_san_solver.py

```diff
@@ -46,6 +48,128 @@ class NeuroSanSolver(Solver):
     Generic solver implementation that uses Neuro SAN.
     """
 
+    def solve(self, problem: str, depth: int, max_depth: int, path: str) -> dict[str, Any]:
+        """
+        Internal recursive solver that returns (response, trace_node).
+        Builds a complete trace tree of the decomposition process.
+
+        :return: The root trace node of the decomposition process
+        """
+        logging.info(f"[solve] depth={depth} path={path} problem: {problem[:120]}{'...' if len(problem) > 120 else ''}")
+
+        node = {
+            "depth": depth,
+            "path": path,
+            "problem": problem,
+            "decomposition": None,
+            "children": [],
+            "sub_finals": None,
+            "composition": None,
+            "response": None,
+            "final": None,
+            "final_num": None,
+            "extracted_final": None,
+            "error": None,
+        }
+
+        if depth >= max_depth:
+            logging.info(f"[solve] depth={depth} -> atomic (max depth)")
+            resp, finals, votes, winner_idx, solutions = self._solve_atomic_with_voting(problem)
+            node["response"] = resp
+            node["final"] = finals[winner_idx]
+            node["atomic"] = {
+                "atomic_candidates": finals,
+                "atomic_votes": votes,
+                "atomic_winner_idx": winner_idx,
+                "final_choice": finals[winner_idx],
+            }
+            node["extracted_final"] = self.extract_final(resp)
+            return node
+
+        p1, p2, c, decomp_meta = self.decompose(problem)
+
+        if not p1 or not p2 or not c:
+            logging.info(f"[solve] depth={depth} -> atomic (no decomp)")
+            if decomp_meta:
+                node["decomposition"] = {**decomp_meta, "decision": "no_decomposition"}
+            resp, finals, votes, winner_idx, solutions = self._solve_atomic_with_voting(problem)
+            node["response"] = resp
+            node["final"] = finals[winner_idx]
+            node["atomic"] = {
+                "atomic_candidates": finals,
+                "atomic_votes": votes,
+                "atomic_winner_idx": winner_idx,
+                "final_choice": finals[winner_idx],
+            }
+            node["extracted_final"] = self.extract_final(resp)
+            return node
+
+        logging.info(f"[solve] depth={depth} using decomposition")
+        node["decomposition"] = decomp_meta
+
+        s1_node = self.solve(p1, depth + 1, max_depth, f"{path}.0")
+        s2_node = self.solve(p2, depth + 1, max_depth, f"{path}.1")
+        node["children"] = [s1_node, s2_node]
+
+        s1: str = s1_node.get("extracted_final")
+        s2: str = s2_node.get("extracted_final")
+        node["sub_finals"] = {"s1_final": s1, "s2_final": s2}
```

Comment:
Previous code had extra calls which got the response and called extract_final().
Now we can expect extract_final() to be in the node data returned from the recursive calls to solve() above, we just get that.